### ISSUE TITLE
Deployment and Database Migrations Timing Research Spike

### ISSUE THREAD

**Issue by adberglund** [project-koku/koku#860]
We want to make sure our migrations are timed and written in such a way that we do not have an incompatible version of the app running after a migration. 

The aim of this research spike is to look at options, see if there is a standard practice, and return to the team with that information.

**Comment by blentz** [comment]
[Bug Triage]

One of the main outputs from this research should be a proposal for how we will release code to production. This proposal should account for horizontally scaling our environment (multiple API front-end pods, etc.).

There should be a meeting scheduled to discuss outcomes of the research.

**Comment by blentz** [comment]
Today we saw an issue in CI that needs to be included in this research.

The Koku pod includes code to automatically run django migrations on startup.

The time it takes to run migrations is at least O(n * x) and worst-case O(n^x) where x is equal to the number of tenants in the DB.

If the time to migrate all tenants is greater than the liveness probe's timeout value, the pod will fail on startup because the migrations delay the pod startup by too much.

This means that we need to configure the automatic migrations to be disabled in (at least) our production config templates. Second, we need to identify a process for running the migrations separately from the koku pod startup.

**Comment by adberglund** [comment]
Talked with @ilya-w some today and his suggestion was to look into using lifecycle-hooks to get our migrations run during the deployment process as a separate pod. This would keep from running each time a new pod scales up. https://docs.openshift.com/container-platform/3.11/dev_guide/deployments/deployment_strategies.html#lifecycle-hooks

**Comment by adberglund** [comment]
@blentz Also here is the cloudigrade template showing what tehy are doing. https://gitlab.com/cloudigrade/shiftigrade/blob/master/ocp/cloudigrade/deployment-configs.yaml#L93

They are using a recreate strategy and mid-hook, but I *think* a pre hook might be good for our rolling strategy.

**Comment by blentz** [comment]
I spent a couple days trying to get this to work, but was unsuccessful.

Here is a branch with my work on implementing a pre-hook: https://github.com/project-koku/koku/tree/predeploy_hook

The main problem I ran into was that the pre-hook container was unable to launch.

The error message I received was:
```
    state:
      waiting:
        message: container koku not found
        reason: CreateContainerConfigError
```

My current theory is that there's something busted in the image lookup logic when it pulls the ImageStream from the internal registry.  The pull succeeds, but then it can't seem to find/access the pulled container for whatever reason.
